ir/instr.cpp

```diff
@@ -2227,6 +2227,88 @@ unique_ptr<Instr> Memcpy::dup(const string &suffix) const {
 }
 
 
+vector<Value*> Strlen::operands() const {
+  return { ptr };
+}
+
+void Strlen::rauw(const Value &what, Value &with) {
+  RAUW(ptr);
+}
+
+void Strlen::print(ostream &os) const {
+  os << getName() << " = strlen " << *ptr;
+}
+
+
+// Given (bid, ofs0), encode an expression that returns the next offset ofs2
+// that has a null character.
+// Returns: <found?, ofs2>
+static pair<expr, expr>
+nextNull(const Memory &m, const expr &bid, const expr &ofs0, unsigned i,
+         const string &nextnull_uf) {
+  auto ofs = ofs0 + expr::mkUInt(i, ofs0.bits());
+
+  if (i == strlen_unroll_cnt) {
+    auto ofs2 = expr::mkUF(nextnull_uf, {bid, ofs}, ofs);
+    Pointer ptr2(m, bid, ofs2);
+    Byte b = m.load(ptr2);
+    // Approximate the situation by assuming that there is no null between
+    // ofs and ofs2
+    return { ptr2.inbounds(false, true) && !b.is_ptr() && !b.is_poison(false) &&
+             b.is_zero(), move(ofs2) };
+  }
+
+  Pointer ptr(m, bid, ofs);
+  Byte b = m.load(ptr);
```

Comment:
why not use the regular load()?
It also helps ensuring that all dereferenciability checks are done properly

Reply:
The result of strlen with unroll_cnt=2 looks conceptually like this:

c1 = load(ptr)
c2 = load(ptr+1)
..
return expr::mkIf(c1 == '\0', 0, mkIf(c2 == '\0', 1, ..))

The second load should check dereferenceability only when c1 != '\0'.
The regular load unconditionally checks dereferenceability, leaving more UBs to the state than required.

Reply:
Maybe load can get an extra parameter to fill in the UB stuff there rather than doing s.addUB()?

Reply:
That sounds good to me. I'll make a separate PR for this.

Reply:
It involves making Pointer::is_dereferenceable do that - do you think it's okay? I think the change seems good because it makes things less stateful.

BTW, the regular load() calls rewriteUndef at the end. I think rewriteUndef should not be called when load() is called from strlen. If unrolled, it will generate huge number of undef quantifiers. What do you think?
```
*p = undef
%a = strlen(p)
%b = strlen(p) // allow %b to be different from %a?
```